Show properties for goto-diff

peterschrammel opened the pull request:
Pre-work:
- Replace cout in show_properties
- Factor out show properties command line def and docs

Show properties for goto-diff:
- Expose conversion of properties for a goto program into a JSON array so that it can be used in goto-diff
- Integrate into goto-diff to show properties for each new/modified/deleted function

Post-work:
- Enable goto check an cover options in goto-diff to have some properties to show
- Use CPROVER exit codes in goto-diff
- Add a test for goto-diff show properties

peterschrammel committed:
Test for goto-diff show properties

peterschrammel committed:
Replace cout in show_properties

All output should be printed to message streams

peterschrammel committed:
Expose conversion of properties of a goto-program into a JSON array

so that it can be reused to goto-diff

peterschrammel committed:
Factor out show properties command line def and docs

to avoid repetition across the CPROVER tools

peterschrammel committed:
Enable goto check and cover options in goto-diff

to be able to compare properties.
The java options are added to be able to use
--java-throw-runtime-exceptions.

peterschrammel committed:
Use CPROVER exit codes

tautschnig commented on src/goto-diff/goto_diff_parse_options.cpp:
```diff
@@ -459,14 +484,16 @@ bool goto_diff_parse_optionst::process_goto_program(
-  catch(int)
+  catch(int e)
   {
+    error() << "Numeric exception : " << e << eom;
```
Nit pick: I don't think there should be a whitespace before `:`?

tautschnig commented on src/goto-diff/goto_diff_parse_options.h:
```diff
@@ -27,6 +30,8 @@ class optionst;
   "(json-ui)" \
   OPT_SHOW_GOTO_FUNCTIONS \
+  OPT_GOTO_CHECK \
+  "(cover):" \
```
Why is it not necessary to add any of the Java stuff that got added to the help output in here?

tautschnig commented on src/goto-diff/goto_diff_base.cpp:
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
     case ui_message_handlert::uit::XML_UI:
     {
-      out << "not supported yet";
+      error() << "not supported yet" << eom;
```
Would you mind changing the error message as well to "XML output not yet supported"?

tautschnig commented on src/goto-diff/goto_diff_base.cpp:
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
-      namespacet(goto_model1.symbol_table).lookup(function_name);
-    result["sourceLocation"] = json(symbol.location);
+    const goto_programt &goto_program = goto_function_it->second.body;
+      for(const auto &ins : goto_program.instructions)
```
The `for` loop seems to be indented more than the  preceding line? Also: may I suggest to move the `const auto goto_function_it = ...` in here as it is not necessary for the other output just above?

tautschnig commented on src/goto-diff/goto_diff_base.cpp:
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
+
+  const goto_programt &goto_program = goto_function_it->second.body;
+  const symbolt &symbol = ns.lookup(function_name);
+  result["sourceLocation"] = json(symbol.location);
```
The order of instructions in this function seems to be almost random.  Here's what I would do:

````c++
const namespacet ns(goto_model.symbol_table);
const symbolt &symbol = ns.lookup(function_name);

result["name"]=json_stringt(id2string(function_name));
result["sourceLocation"] = json(symbol.location);

if(options.get_bool_option("show-properties"))
{
  const auto goto_function_it = goto_model.goto_functions.function_map.find(function_name);
  ...
````

tautschnig commented on src/goto-diff/goto_diff_parse_options.cpp:
```diff
@@ -234,6 +236,12 @@ void goto_diff_parse_optionst::get_command_line_options(optionst &options)
+    options.set_option("show-properties", true);
+  else
+    options.set_option("show-properties", false);
+
```
`options.set_option("show-properties", cmdline.isset("show-properties"));` and no need for blank line before closing `}`

peterschrammel committed:
Show properties in goto-diff

Shows properties for each new/modified/deleted function.
For now, for modified functions all properties are shown
regardless of whether they are affected by the
modification or not.

chrisr-diffblue reviewed (approved):
Looks ok to me.
